File: examples/content_identity/verify_golden.mjs
// Verify the portable M2RIV v1 content-identity profile using Node stdlib only.
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";

const vectorsUrl = new URL("./golden-vectors.json", import.meta.url);
const document = JSON.parse(readFileSync(vectorsUrl, "utf8"));

function compareUnicodeScalar(left, right) {
  const a = Array.from(left, (character) => character.codePointAt(0));
  const b = Array.from(right, (character) => character.codePointAt(0));
  for (let index = 0; index < Math.min(a.length, b.length); index += 1) {
    if (a[index] !== b[index]) return a[index] - b[index];
  }
  return a.length - b.length;
}

function canonicalJson(value) {
  if (value === null || typeof value === "boolean" || typeof value === "string") {
    return JSON.stringify(value);
  }
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new TypeError("numbers must be finite");
    if (Object.is(value, -0)) return "-0.0";
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(canonicalJson).join(",")}]`;
  }
  if (typeof value === "object") {
    const keys = Object.keys(value).sort(compareUnicodeScalar);
    return `{${keys
      .map((key) => `${JSON.stringify(key)}:${canonicalJson(value[key])}`)
      .join(",")}}`;
  }
  throw new TypeError(`unsupported value: ${typeof value}`);
}

for (const vector of document.vectors) {
  const payload = canonicalJson(vector.value);
  if (payload !== vector.canonical_json) {
    throw new Error(`canonical JSON mismatch: ${vector.name}`);
  }
  const domain = Buffer.from(`m2riv:${vector.namespace}:v1`, "utf8");
  const digest = createHash("sha256")
    .update(domain)
    .update(Buffer.from([0]))
    .update(Buffer.from(payload, "utf8"))
    .digest("hex");
  if (digest !== vector.sha256) {
    throw new Error(`fingerprint mismatch: ${vector.name}`);
  }
}

console.log(`verified ${document.vectors.length} M2RIV v1 identity vectors`);

File: examples/content_identity/verify_golden.py
"""Verify the public MCR v1 content-identity vectors using only stdlib."""

from __future__ import annotations

import hashlib
import json
import struct
from collections.abc import Mapping
from datetime import datetime
from pathlib import Path
from typing import Any

from typed_values import materialize_typed_value

VECTORS = Path(__file__).with_name("golden-vectors.json")
FLOAT_VECTORS = Path(__file__).with_name("float-vectors.json")


def jsonable(value: Any) -> Any:
    if isinstance(value, Mapping):
        return {key: jsonable(item) for key, item in value.items()}
    if isinstance(value, (list, tuple)):
        return [jsonable(item) for item in value]
    if isinstance(value, (set, frozenset)):
        return sorted((jsonable(item) for item in value), key=repr)
    if isinstance(value, datetime):
        return value.isoformat()
    if isinstance(value, Path):
        return value.as_posix()
    return value


def canonical_json(value: Any) -> bytes:
    return json.dumps(
        jsonable(value),
        ensure_ascii=False,
        allow_nan=False,
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")


def fingerprint(value: Any, namespace: str) -> str:
    domain = f"mcr:{namespace}:v1".encode()
    return hashlib.sha256(domain + b"\0" + canonical_json(value)).hexdigest()


def main() -> None:
    document = json.loads(VECTORS.read_text(encoding="utf-8"))
    for vector in document["vectors"]:
        source = vector.get("typed_value", vector.get("value"))
        value = materialize_typed_value(source)
        actual_json = canonical_json(value).decode()
        if actual_json != vector["canonical_json"]:
            raise SystemExit(f"canonical JSON mismatch: {vector['name']}")
        if fingerprint(value, vector["namespace"]) != vector["sha256"]:
            raise SystemExit(f"fingerprint mismatch: {vector['name']}")
    print(f"verified {len(document['vectors'])} MCR v1 identity vectors")

    float_document = json.loads(FLOAT_VECTORS.read_text(encoding="utf-8"))
    for vector in float_document["vectors"]:
        value = struct.unpack(">d", bytes.fromhex(vector["bits"]))[0]
        payload = {"value": value}
        if canonical_json(payload).decode() != vector["canonical_json"]:
            raise SystemExit(f"float canonical JSON mismatch: {vector['bits']}")
        if fingerprint(payload, "float-spelling-corpus") != vector["sha256"]:
            raise SystemExit(f"float fingerprint mismatch: {vector['bits']}")
    print(f"verified {len(float_document['vectors'])} binary64 spelling vectors")


if __name__ == "__main__":
    main()
